File: results/0.0/Bitcoin-transactions/bitcoin-transactions-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Parser for a variable length integer (VarInt)
static HParser *varint() {
    return h_choice(h_int64(), h_int32(), h_int16(), h_int8(), NULL);
}

// Parser for a transaction input
static HParser *tx_input() {
    HParser *prev_tx_hash = h_bits(256, false);
    HParser *prev_tx_index = h_bits(32, false);
    HParser *script_length = varint();
    HParser *script_sig = h_length_value(script_length, h_uint8());
    HParser *sequence = h_bits(32, false);

    return h_sequence(prev_tx_hash, prev_tx_index, script_length, script_sig, sequence, NULL);
}

// Parser for a transaction output
static HParser *tx_output() {
    HParser *value = h_bits(64, false);
    HParser *script_length = varint();
    HParser *script_pubkey = h_length_value(script_length, h_uint8());

    return h_sequence(value, script_length, script_pubkey, NULL);
}

// Parser for a Bitcoin transaction
static HParser *bitcoin_tx() {
    HParser *version = h_bits(32, false);
    HParser *input_count = varint();
    HParser *inputs = h_many(tx_input());
    HParser *output_count = varint();
    HParser *outputs = h_many(tx_output());
    HParser *lock_time = h_bits(32, false);

    return h_sequence(version, input_count, inputs, output_count, outputs, lock_time, NULL);
}

// Main function to parse a Bitcoin transaction from a binary file
int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <binary transaction file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    // Seek to the end of the file to determine the file size
    fseek(file, 0, SEEK_END);
    size_t length = ftell(file);
    fseek(file, 0, SEEK_SET);

    // Read the entire file into a buffer
    uint8_t *buffer = malloc(length);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    if (fread(buffer, 1, length, file) != length) {
        perror("Failed to read file");
        free(buffer);
        fclose(file);
        return 1;
    }

    fclose(file);

    // Create a parser for a Bitcoin transaction
    HParser *parser = bitcoin_tx();
    HParseResult *result = h_parse(parser, buffer, length);

    if (result) {
        printf("Transaction parsed successfully.\n");
        h_pprint(stdout, result->ast, 0, 0);
    } else {
        printf("Failed to parse transaction.\n");
    }

    // Clean up
    h_parse_result_free(result);
    h_free_parser(parser);
    free(buffer);

    return 0;
}

File: results/0.0/Bitcoin-transactions/bitcoin-transactions-gpt-4-turbo.rs
use nom::{
    bytes::complete::take,
    combinator::map,
    multi::count,
    number::complete::{le_i32, le_u32, le_u64},
    sequence::tuple,
    IResult,
};
use std::fs::File;
use std::io::{self, Read};
use std::env;

#[derive(Debug)]
struct Transaction {
    version: i32,
    inputs: Vec<TxInput>,
    outputs: Vec<TxOutput>,
    lock_time: u32,
}

#[derive(Debug)]
struct TxInput {
    previous_output: TxOutPoint,
    script_sig: Vec<u8>,
    sequence: u32,
}

#[derive(Debug)]
struct TxOutPoint {
    hash: Vec<u8>,
    index: u32,
}

#[derive(Debug)]
struct TxOutput {
    value: u64,
    script_pub_key: Vec<u8>,
}

fn parse_varint(input: &[u8]) -> IResult<&[u8], (u64, &[u8])> {
    let first_byte = input[0];
    match first_byte {
        0xfd => map(take(2usize), |bytes: &[u8]| (u64::from_le_bytes([bytes[0], bytes[1], 0, 0, 0, 0, 0, 0]), &input[3..]))(input),
        0xfe => map(take(4usize), |bytes: &[u8]| (u64::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3], 0, 0, 0, 0]), &input[5..]))(input),
        0xff => map(take(8usize), |bytes: &[u8]| (u64::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7]]), &input[9..]))(input),
        _ => Ok((&input[1..], (first_byte as u64, &input[1..])),
    }
}

fn parse_input(input: &[u8]) -> IResult<&[u8], TxInput> {
    let (input, previous_output) = parse_outpoint(input)?;
    let (input, (script_length, input)) = parse_varint(input)?;
    let (input, script_sig) = take(script_length)(input)?;
    let (input, sequence) = le_u32(input)?;
    Ok((input, TxInput {
        previous_output,
        script_sig: script_sig.to_vec(),
        sequence,
    }))
}

fn parse_outpoint(input: &[u8]) -> IResult<&[u8], TxOutPoint> {
    let (input, hash) = take(32usize)(input)?;
    let (input, index) = le_u32(input)?;
    Ok((input, TxOutPoint {
        hash: hash.to_vec(),
        index,
    }))
}

fn parse_output(input: &[u8]) -> IResult<&[u8], TxOutput> {
    let (input, value) = le_u64(input)?;
    let (input, (script_length, input)) = parse_varint(input)?;
    let (input, script_pub_key) = take(script_length)(input)?;
    Ok((input, TxOutput {
        value,
        script_pub_key: script_pub_key.to_vec(),
    }))
}

fn parse_transaction(input: &[u8]) -> IResult<&[u8], Transaction> {
    let (input, version) = le_i32(input)?;
    let (input, (input_count, input)) = parse_varint(input)?;
    let (input, inputs) = count(parse_input, input_count as usize)(input)?;
    let (input, (output_count, input)) = parse_varint(input)?;
    let (input, outputs) = count(parse_output, output_count as usize)(input)?;
    let (input, lock_time) = le_u32(input)?;
    Ok((input, Transaction {
        version,
        inputs,
        outputs,
        lock_time,
    }))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <file_path>", args[0]);
        return Ok(());
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_transaction(&buffer) {
        Ok((_, transaction)) => println!("{:#?}", transaction),
        Err(e) => eprintln!("Failed to parse transaction: {:?}", e),
    }

    Ok(())
}